### DISCUSSION TITLE
BLE packet loss

### DISCUSSION THREAD

**Discussion by leydar** [capacitor-community/bluetooth-le#422] [Q&A] [answered]
The BLE protocol is lossless. As in, it will drop the connection in the event of packet loss (where the resent packet has also been dropped). Yet when I move a device out of range or otherwise occlude it from the app, sequential packets are lost. The connection isn't dropped but there are gaps in the data transfer. Is it possible that these packets are received at the protocol level but don't make it through to the notifications callback?

My code for testing is straightforward and I'm using v2.0.1.
```
    await BleClient.initialize({ androidNeverForLocation: true });
    await BleClient.connect(macAddress, onDisconnect);
    await BleClient.startNotifications(macAddress, service, characteristic, cb);
```

**Comment by peitschie** [comment]
Just as a point of clarity, Bluetooth **notifications** explicitly may be dropped without penalty. Bluetooth **indications** require an ACK, and would force the connection to drop if this packet was lost: https://www.lairdconnect.com/support/faqs/what-difference-between-indication-and-notification

I'm not entirely sure which you are talking about specifically given the terms of often used interchangeably 🙂 

If the characteristic you are subscribing to supports both notifications AND indications, then notifications will be taken in preference based on https://github.com/capacitor-community/bluetooth-le/blob/main/android/src/main/java/com/capacitorjs/community/plugins/bluetoothle/Device.kt#L400-L406 (assuming you are dealing with Android here).

**Comment by leydar** [comment]
Thank you for your reply. 

My understanding of notifications is that they have a Link-Layer ack and preserve the BLE lossless behaviour. The ack required by an indication is at the application layer. We are using notifications as the time overhead for indication was too high.

**Reply by peitschie** [reply]
Interesting! What you say about the link layer makes sense.

Personally, I've definitely seen notifications (not indications) vanish in transit without the connection dropping. But... I'm not sure if this was due to the peripheral not sending (e.g., due to buffering too many packets at a hardware level), or the phone dropping them for some reason.

**Reply by leydar** [reply]
We just did a test here for completeness. Switching from notifications to indications caused way more packet loss at the application level. Like every second and third message has been dropped with the connection remaining unaffected. It does look like the messages aren't being surfaced quickly enough by the library.

**Comment by pwespi** [comment, marked as answer]
https://github.com/capacitor-community/bluetooth-le/issues/424#issuecomment-1286785382 :

> > Does it work correctly with apps like nordic's nrfconnect? do you receive every notification?
> > Is it purely an issue with this plugin? If so, did changing the timing affect it? try at say 500ms and work your way down.
> 
> I recorded the UI in LightBlue and tried to watch it frame by frame but figured the UI might be throttled. nRF Connect has logging so I was able to see missed packets there as well. Thank you for the suggestion.
> 
> It looks like the issue is at the sender rather than with the library. Sorry for the misdirection.

Thank you all for the discussion. Looks like it turned out to be an issue on the sender and not with the library.
